pkg/reference/ref.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reference
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
```

Comment:
Not sure what the answer is here, but should we move this into `apis/v1alpha2/utils` to correspond with the validation package? I think that would make sense if we wanted to provide a unique set of utils for each API version. Alternatively, we may want to just tie this package to our latest API version, which would make the current placement make sense. I think it's likely sufficient to just support the latest API version here, but if that's the case, we should be clear about that goal up front.

Reply:
Yeh I thought about that. We already have unversioned packages that import v1alpha2 though, so if we do that we should be consistent.

Reply:
I tried putting the generic code in `pkg/reference` and the API-specific code in `apis/v1alpha2/reference`. The use of 2 "reference" packages is awkward but don't really have better names. I immediately ran into circular import problems in the tests which sucked.

Tabling this for today. I'll think it over and try to update the PR later next week.